conanfile.py

```diff
@@ -111,6 +115,9 @@ def _configure_cmake(self):
             cmake.definitions["CMAKE_CXX_FLAGS"] = "-D_WIN32_WINNT=0x600"
             cmake.definitions["CMAKE_C_FLAGS"] = "-D_WIN32_WINNT=0x600"
 
+        if self.settings.os == "Windows" and self.settings.compiler == "Visual Studio":
+            cmake.definitions["Protobuf_USE_STATIC_LIBS"] = "ON"
```

Comment:
I don't understand why this is necessary. Also on a very quick look I could not found this definition used in grpc. Could you point me in the right direction?

Also people can use a shared version of protobuf by setting `protobuf:shared=True`. Setting this flag should reflect the actual value

Reply:
I will try to explain.

`grpc/CMakeLists.txt` includes `grpc/cmake/protobuf.cmake` to find protobuf.
There's a variable in `grpc/CMakeLists.txt` to control which way to invoke `find_package` (Module Mode or Config Mode). See line 70 in `grpc/CMakeLists.txt`:

```cmake
set(gRPC_PROTOBUF_PACKAGE_TYPE "" CACHE STRING "Algorithm for searching protobuf package")
set_property(CACHE gRPC_PROTOBUF_PACKAGE_TYPE PROPERTY STRINGS "CONFIG" "MODULE")
```

In default this value is empty, which makes `grpc/cmake/protobuf.cmake` to invoke `find_package` in basic signature, see `grpc/cmake/protobuf.cmake` line 55:

```cmake
  find_package(Protobuf REQUIRED ${gRPC_PROTOBUF_PACKAGE_TYPE})
```

Which's equaling to:
```cmake
find_package(Protobuf REQUIRED)
```

According to doc of CMake, in this basic signature, I think CMake is going to find `FindProtobuf.cmake` first,  if it fails, then CMake will fallback to Config mode, which is trying to find `ProtobufConfig.cmake`.

Eventually, CMake find a `FindProtobuf.cmake` which comes from CMake itself, for Windows, is located in `C:/Program Files/CMake/share/cmake-3.17/Modules/FindProtobuf.cmake`.

Since CMake 3.17 there's change in this cmake modules, you can check this [commit](https://github.com/Kitware/CMake/commit/3bbd85d5fffe66181cf16c81b23b2ba50f5387ba) and this [commit](https://github.com/Kitware/CMake/commit/a6cc72595f01b517bc54b8b745c2eacf2a38c160). It defines `PROTOBUF_USE_DLLS` in default for MSVC, I think this cause CI of Windows goes failed recent months.

We're linking `protobuf@bincrafters` in static by default, but this cmake modules assumed we're linking in dynamic, that why linking errors occurred during compiling Windows version.

There're several ways to fix this I can figure out:
1. Set `Protobuf_USE_STATIC_LIBS` to prevent `FindProtobuf.cmake` adding definition to `protobuf::libprotobuf`.
2. Set `gRPC_PROTOBUF_PACKAGE_TYPE` to "CONFIG" to make CMake use `protobufConfig.cmake`, which is generated by conan `cmake_find_package_multi` generator, may be a solution.
3. Use dynamic library of protobuf for Windows, but I won't seriously consider this as a solution, because during my investigation, I found protobuf don't recommend using dynamic library on Windows, see discuss [here](https://github.com/protocolbuffers/protobuf/issues/2219#issuecomment-251510320), besides that, `protobuf@bincrafters` didn't pack dynamic binaries for Windows.

I'm trying the first way. But, yes, as you point out, if user trying to use `protobuf:shared=True` on Windows, we should not define `Protobuf_USE_STATIC_LIBS`.

But I'm not sophisticated in creating conan package in python, I'll try to fix situation `protobuf:shared=True` on Windows and get test, I'll be back.